include/mata/sparse-set.hh

```diff
@@ -0,0 +1,355 @@
+//
+// Created by Lukáš Holík on 10.06.2023.
+// Based on
+/**
+	SparseSet.h
+	Implements a class template of a sparse
+	set of unsigned integers.
+	@author Sam Griffiths
+	www.computist.xyz
+    https://gist.github.com/sjgriffiths/06732c6076b9db8a7cf4dfe3a7aed43a
+*/
+//there may be better implementations out there, something in boost, or
+//https://github.com/lluchs/sparsearray
+//https://stackoverflow.com/questions/76205534/is-there-high-performance-data-structure-for-sparse-bit-map-set-semantic-queries
+//https://gist.github.com/Gpinchon/19dd224d5e7fc92596e01f5df6508e2a
+//https://www.codeproject.com/Articles/859324/Fast-Implementations-of-Sparse-Sets-in-Cplusplus
+
+#ifndef LIBMATA_SPARSE_SET_HH
+#define LIBMATA_SPARSE_SET_HH
+
+#include <cassert>
+#include <vector>
+#include <type_traits>
+#include <mata/ord-vector.hh>
+
+namespace Mata::Util {
+
+/**
+ * @brief  Implementation of a set of non-negative numbers using sparse-set.
+ *
+ *
+ * This class implements the interface of a set (similar to std::set) using sparse-set date structure,
+ * that is, a pair of vectors dense and sparse (... google it).
+ * Importantly
+ * - Insertion and removal are constant time.
+ * - Iteration is linear in the number of stored elements.
+ * - It takes a lot of space, the sparse and dense vectors allocate as many indexes as the maximal stored number.
+ *
+ * @tparam  Number  Number type: type of numbers contained in the container.
+ */
+    template<typename Number>
+    class SparseSet {
+        static_assert(std::is_unsigned<Number>::value, "SparseSet can only contain unsigned integers");
+
+    private:
+        std::vector<Number> dense;    //Dense set of elements
+        std::vector<Number> sparse;    //Map of elements to dense set indices
+
+        size_t size_ = 0;    //Current size (number of elements)
+        size_t capacity_ = 0;    //Current capacity (maximum value + 1 or more (can be more when maximum is removed))
```

Comment:
As discussed, rename to `domain_size`.

Capacity can be confused with the C++ normal capacity of STL containers.

Reply:
done